mmpose/models/heads/heatmap_heads/multi_stage_heatmap_head.py

```diff
@@ -0,0 +1,643 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import copy
+from typing import List, Optional, Sequence, Union
+
+import torch
+from mmcv.cnn import (ConvModule, DepthwiseSeparableConvModule, Linear,
+                      build_activation_layer, build_conv_layer,
+                      build_norm_layer, build_upsample_layer)
+from mmengine.data import PixelData
+from torch import Tensor, nn
+
+from mmpose.core.utils.tensor_utils import to_numpy
+from mmpose.core.utils.typing import (ConfigType, MultiConfig, OptConfigType,
+                                      OptSampleList, SampleList)
+from mmpose.metrics.utils import pose_pck_accuracy
+from mmpose.registry import KEYPOINT_CODECS, MODELS
+from ..base_head import BaseHead
+
+OptIntSeq = Optional[Sequence[int]]
+
+
+@MODELS.register_module()
+class MultiStageHeatmapHead(BaseHead):
+    """Multi-stage heatmap head introduced in `Convolutional Pose Machines`_
+    and `Stacked Hourglass Networks`_. It consists of multiple branches, each
+    of which has some deconv layers and a simple conv2d layer.
+
+    Args:
+        num_stages (int): Number of stages.
+        in_channels (int | Sequence[int]): Number of channels in the input
+            feature maps.
+        out_channels (int): Number of channels in the output heatmaps.
+        deconv_out_channels (Sequence[int], optional): The output channel
+            number of each deconv layer. Defaults to ``(256, 256, 256)``
+        deconv_kernel_sizes (Sequence[int | tuple], optional): The kernel size
+            of each deconv layer. Each element should be either an integer for
+            both height and width dimensions, or a tuple of two integers for
+            the height and the width dimension respectively.
+            Defaults to ``(4, 4, 4)``
+        has_final_layer (bool): Whether have the final 1x1 Conv2d layer.
+            Defaults to ``True``
+        loss (Config | List[Config]): Config of the keypoint loss of different
+            stages. Defaults to use :class:`KeypointMSELoss`.
+        decoder (Config, optional): The decoder config that controls decoding
+            keypoint coordinates from the network output. Defaults to ``None``
+        init_cfg (Config, optional): Config to control the initialization. See
+            :attr:`default_init_cfg` for default settings
+
+    .. _`Convolutional Pose Machines`: https://arxiv.org/abs/1602.00134
+    .. _`Stacked Hourglass Networks`: https://arxiv.org/abs/1603.06937
+    """
+
+    _version = 2
+
+    def __init__(self,
+                 num_stages: int,
+                 in_channels: Union[int, Sequence[int]],
```

Comment:
I suggest keep `in_channes` and `out_channels` as the first 2 parameters.

Reply:
Done